src/can.rs

```diff
@@ -0,0 +1,145 @@
+//! # Controller Area Network (CAN) Interface
+//!
+
+use crate::bb;
+#[cfg(any(feature = "stm32f405", feature = "stm32f407"))]
+use crate::gpio::{
+    gpioa::{PA11, PA12},
+    gpiob::{PB12, PB13, PB5, PB6, PB8, PB9},
+    gpiod::{PD0, PD1},
+    gpioh::PH13,
+    gpioi::PI9,
+    Alternate, AF9,
+};
+
+#[cfg(feature = "stm32f446")]
+use crate::gpio::{
+    gpioa::{PA11, PA12},
+    gpiob::{PB12, PB13, PB5, PB6, PB8, PB9},
+    gpiod::{PD0, PD1},
+    Alternate, AF9,
+};
+use crate::pac::{CAN1, CAN2};
+use crate::rcc::APB1;
+
+mod sealed {
+    pub trait Sealed {}
+
+    /// Bus associated to peripheral
+    pub trait RccBus {
+        /// Bus type;
+        type Bus;
+    }
+}
+
+pub trait Pins: sealed::Sealed {
+    type Instance;
+}
+
+/*
+    order: tx, rx similar to serial
+*/
+macro_rules! pins {
+    ($($PER:ident => ($tx:ident, $rx:ident),)+) => {
+        $(
+            impl sealed::Sealed for ($tx<Alternate<AF9>>, $rx<Alternate<AF9>>) {}
+            impl Pins for ($tx<Alternate<AF9>>, $rx<Alternate<AF9>>) {
+                type Instance = $PER;
+            }
+        )+
+    }
+}
+
+/*
+    See DS8626 Rev 9 Table 9.
+*/
+#[cfg(any(feature = "stm32f405", feature = "stm32f407"))]
+pins! {
+    CAN1 => (PA12, PA11),
+    CAN1 => (PB9, PB8),
+    CAN1 => (PD1, PD0),
+    CAN1 => (PH13, PI9),
+    CAN2 => (PB13, PB12),
+    CAN2 => (PB6, PB5),
+}
+
+/*
+    See DS10693 Rev 9 Table 11.
+*/
+#[cfg(feature = "stm32f446")]
+pins! {
+    CAN1 => (PA12, PA11),
+    CAN1 => (PB9, PB8),
+    CAN1 => (PD1, PD0),
+    CAN2 => (PB13, PB12),
+    CAN2 => (PB6, PB5),
+}
+
+use sealed::RccBus;
+
+/// Enable/disable peripheral
+pub trait Enable: RccBus {
+    fn enable(apb: &mut Self::Bus);
+}
+
+macro_rules! bus {
+    ($($PER:ident => ($apbX:ty, $peren:literal),)+) => {
+        $(
+            impl RccBus for crate::pac::$PER {
+                type Bus = $apbX;
+            }
+            impl Enable for crate::pac::$PER {
+                #[inline(always)]
+                fn enable(apb: &mut Self::Bus) {
+                    unsafe { bb::set(apb.enr(), $peren) };
+                }
+            }
+        )+
+    }
+}
+
+bus! {
+    CAN1 => (APB1, 25),
+    CAN2 => (APB1, 26),
+}
+
+/// Interface to the CAN peripheral.
+pub struct Can<Instance> {
+    _peripheral: Instance,
+}
+
+impl<Instance> Can<Instance>
+where
+    Instance: Enable<Bus = APB1>,
+{
+    /// Creates a CAN interaface.
+    pub fn new<P>(can: Instance, _pins: P, apb: &mut APB1) -> Can<Instance>
+    where
+        P: Pins<Instance = Instance>,
+    {
+        Instance::enable(apb);
```

Comment:
I would prefer the use of bit banding here. It might be good to also reset the peripheral here (with bit banding too)